### ISSUE TITLE
JS Testing Feature: Multiple return values from interaction

### ISSUE THREAD

**Issue by srinjoyc** [onflow/flip-fest#30]
> 👋  &nbsp; If you are interested in working on this issue, please check out the [Getting Started guide on HackerEarth](https://www.hackerearth.com/challenges/hackathon/flip-fest/custom-tab/getting-started/#Getting%20Started)!

## Description (Problem Statement)

The JS Testing Framework has two functions: `sendTransaction` and `executeScript` that currently return a single result from their execution. It does not have a good way of handling errors that should be caught and processed. This can cause issues with multiple interactions called in succession.

## Experience Required

- Great proficiency with JavaScript
- Good understanding of Cadence and Flow JS SDK

## Minimum Feature Set (Acceptance Criteria)

- Catch error inside interaction code and always return tuple `[result, error]`.

- This should make it easier to assert the result of interaction using jest and other libraries.

- Flow Cadut interactions are already implemented in such a way (refer to this piece of code as an example - [https://github.com/onflow/flow-cadut/blob/main/src/interactions.js#L55-L63](https://github.com/onflow/flow-cadut/blob/main/src/interactions.js#L55-L63)) so it's mostly a refactoring task.

## Extension Feature Set (Optional)

Parse error message and provide an error in form of an object, containing at least `errorCode`, `message`. If possible extract more information and format it in a human-readable and usable way.

## Milestone Requirements

1. Meet the minimum feature set and optionally the extension.

## Software Requirements

*Maintainability*

Code should be written with best practices in mind.

## Other Requirements

*Documentation*

- Documentation should be updated according to the changes made.

*Code standards or guidelines*

- Follow the general [contribution guidelines](https://github.com/onflow/flow-js-testing/blob/master/CONTRIBUTING.md).

## Judging Criteria

- For all projects, [technical & high-level judging criteria apply](https://github.com/onflow/flip-fest/blob/main/resources/judging-criteria.md).

## Resources

- [General Resource Guide](https://docs.google.com/document/d/1VXStCyTYRjGNCzTBUW9JAgVenfvLghKq3iMGkMlmroU/edit#heading=h.qpn9u6p0l5y7)
- Flow JS Testing
    - [Repository](https://github.com/onflow/flow-js-testing)
    - [Flow Cadut](https://github.com/onflow/flow-cadut)

**Comment by MaxStalker** [comment]
Hey, Flow Surfer! 
My name is Max. I am core member of Developer Experience team and I will be your Point-of-Contact for this task.
If you stumble into any challenges with this task, feel free to contact me via replies to this issue or over [Flow Discord](https://discord.gg/86h3jGW8) (my handle is `MaxStarka#5714`).

Looking forward to work with you! :)

**Comment by OmarMalik** [comment]
Hey Max, I want to make sure I'm understanding this correctly.

Since we are changing the return types of `executeScript `and `executeTransaction `in the `flow-js-testing` library, won't that break any existing code that uses these functions? Not sure how to get around this.

For example, I'm imagining the intended refactor for executeScript is the following:
```
export const executeScript = async (...props) => {
    try {
        const extractor = extractParameters("script");
        const { code, args } = await extractor(props);
        
        const ix = [fcl.script(code)];
        // add arguments if any
        if (args) {
          ix.push(fcl.args(resolveArguments(args, code)));
        }
        const response = await fcl.send(ix);
        const result = await fcl.decode(response);
        [result, null]
    } catch (e) {
        return [null, e]
    }
}

```

Am I on the right track here?

**Comment by MaxStalker** [comment]
@OmarMalik that is correct - this will be breaking change if your code previously wasn't using Jest helper methods.

There are 3 ways to go:
- use `fcl.config` param to decide how returned value shall look and whether to catch error or not
- introduce the breaking change with new release and make it clear how developer can refactor it
- mix two aforementioned solutions by introducing new `fcl.config`param in next release and deprecate it and old way of doing things in following

Your call ;)

**Comment by MaxStalker** [comment]
Hey, @OmarMalik are you still working on this issue? :)

**Comment by hichana** [comment]
@OmarMalik if you are still working on this FLIP issue, can you let us know? If not, and if ok with @MaxStalker I'm hoping to do this one or another FLIP in the time that remains with @j00lz. Thanks

**Comment by MaxStalker** [comment]
おめでとう, @hichana ！=^____^=
🎉 **Milestone Achieved** 🎉

**Comment by MaxStalker** [comment]
🎉 **Milestone Achieved!** 🎉 

Thank you for your submission and participation!

**Comment by hichana** [comment]
Cheers @MaxStalker :)

**Comment by kimcodeashian** [comment]
Good day @hichana!

Thanks so much for all your hardwork & participation. In order to finalize winners & prepare for prize payout, we'll need the following actions from your end.

Please provide the following information by **Nov 17, 2021**,  (in this GH Issue is fine): 

**1. Team Information**
  - Team Members Information - Github Username + Email Contact + Percentage of prize allocation (total should = 100%) 
  - All mentioned members MUST react to the post with a 👍 which will act as confirmation that the information is correct, or a 👎 to indicate that the information is not correct.
  - We will be reaching out via e-mail

**🎖IMPORTANT**: We will only proceed with prize payouts once all members have confirmed with 👍 on the post. 


**2. Video Demo (optional)**
  - Please provide a 5-minute video demo to be featured & showcased in the FLIP Fest Closing Ceremonies
  - Link format & Downloadable (eg. Google Drive, Vimeo)
  - Content Format (Problem Statement, your work / how you solved it, final outcome) 
 

We will be hosting Closing Ceremonies on November 23rd, 8AM PT where we'll having closing remarks from Dete & will be announcing the winners! I'll share the details here before Nov 17.

**Comment by hichana** [comment]
Thank you :)

Team Information:
- Github Username: [hichana](https://github.com/hichana) and [justjoolz](https://github.com/justjoolz)
- Email Contact: hichana@gmail.com, justjoolz@gmail.com
- Percentage of prize allocation: hichana 50%, justjoolz 50%

**Comment by kimcodeashian** [comment]
Hey folks,

We've received and reviewed over **82 submissions**! What an amazing community on Flow! To commemorate all the hard work done, we have finalized winners and will be announcing them during our **[Closing Ceremony](https://forms.gle/j9REJJkVtFFibtfR8) on Nov 23rd, 8AM PT**. Be sure to join us - there may be some attendance prizes & a keynote from our CTO, Dete 😉!  

**[RSVP here](https://forms.gle/j9REJJkVtFFibtfR8)** so you don't miss out! See you then!
